src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMessage.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Formats.Cbor;
+using System.Runtime.Versioning;
+
+namespace System.Security.Cryptography.Cose
+{
+    public abstract class CoseMessage
+    {
+        private const byte EmptyStringByte = 0xa0;
+        // COSE tags https://datatracker.ietf.org/doc/html/rfc8152#page-8 Table 1.
+        internal const CborTag Sign1Tag = (CborTag)18;
+
+        internal byte[]? _content;
+        internal byte[] _signature;
+        internal byte[] _protectedHeaderAsBstr;
+
+        private CoseHeaderMap _protectedHeaders;
+        private CoseHeaderMap _unprotectedHeaders;
+        public CoseHeaderMap ProtectedHeader => _protectedHeaders;
+        public CoseHeaderMap UnprotectedHeader => _unprotectedHeaders;
+
+        internal CoseMessage(CoseHeaderMap protectedHeader, CoseHeaderMap unprotectedHeader, byte[]? content, byte[] signature, byte[] encodedProtectedHeader)
+        {
+            _content = content;
+            _signature = signature;
+            _protectedHeaderAsBstr = encodedProtectedHeader;
+            _protectedHeaders = protectedHeader;
+            _unprotectedHeaders = unprotectedHeader;
+        }
+
+        // Sign and MAC also refer to the content as payload.
+        // Encrypt also refers to the content as cyphertext.
+        public ReadOnlyMemory<byte>? Content
+        {
+            get
+            {
+                if (_content  != null)
+                {
+                    return _content;
+                }
+
+                return null;
+            }
+        }
```

Comment:
This can just be:
```suggestion
        public ReadOnlyMemory<byte>? Content => _content;
```
It's a bit odd from an API perspective to have a nullable `ReadOnlyMemory<byte>`, though... is that the intention?

Reply:
With your suggested syntax, the implicit operator will kick-in and the property will never return null.

Reply:
> It's a bit odd from an API perspective to have a nullable ReadOnlyMemory<byte>, though... is that the intention?

it is, specially to signify a message that explicitly excluded the content (detached content as is called in the RFC). 

Also, I found some prior art in ou APIs:
https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.sendpacketselement.memorybuffer?view=net-6.0
https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.pkcs.pkcs8privatekeyinfo.algorithmparameters?view=dotnet-plat-ext-6.0

Reply:
Nullable is correct; because there's a distinction between "had empty content" (attached-empty) and "had no content" (detached), which needs to be conveyed to the caller (as it changes which Verify method they need to use)